src/connection.c

```diff
@@ -154,12 +155,41 @@ mpd_connection_new(const char *host, unsigned port, unsigned timeout_ms)
 						&connection->timeout,
 						&connection->error);
 		}
+	}
 #endif
 
-		if (fd == MPD_INVALID_SOCKET)
+	if (fd == MPD_INVALID_SOCKET) {
+		/* special case: try socket in XDG_RUNTIME_DIR */
+		mpd_settings_free(settings);
+		const char *xdg_runtime_dir = getenv("XDG_RUNTIME_DIR");
+		if (xdg_runtime_dir == NULL) {
+			return connection;
+		}
+		size_t xdg_socket_len = strlen(xdg_runtime_dir) + sizeof("/mpd/socket") + 1;
+		char *xdg_socket = malloc(xdg_socket_len);
+		if (xdg_socket == NULL)
+			return connection;
+		snprintf(xdg_socket, xdg_socket_len, "%s%s", xdg_runtime_dir, "/mpd/socket");
```

Comment:
Unnecessary use of stdio, and the buffer size checks are just as unnecessary. We already verified that the allocation has the right size.